pkg/controllers/deploycontroller.go

```diff
@@ -0,0 +1,245 @@
+	utilruntime.Must(clusterv1.AddToScheme(scheme))
+}
+
+const (
+	imageName              = "cluster_proxy_addon"
+	defaultImagePullPolicy = "IfNotPresent"
+	imagePullSecret        = "open-cluster-management-image-pull-credentials"
+
+	// annotationNodeSelector is key name of nodeSelector annotation synced from mch
+	annotationNodeSelector = "open-cluster-management/nodeSelector"
+)
+
+var (
+	agentImage string
+)
+
+func addFlagsForDeployController(cmd *cobra.Command) {
+	cmd.Flags().StringVar(&agentImage, "agent-image", "", "The image of agent")
+}
+
+type reconcileDeployAgentManifestwork struct {
+	client        client.Client
+	addonLister   addonv1alpha1lister.ManagedClusterAddOnLister
+	clusterLister clusterv1lister.ManagedClusterLister
+	workClient    workv1client.Interface
+}
+
+func registerDeployController(addonLister addonv1alpha1lister.ManagedClusterAddOnLister, clusterLister clusterv1lister.ManagedClusterLister, workClient workv1client.Interface, mgr manager.Manager) error {
+	c, err := controller.New("deploy-controller", mgr, controller.Options{
+		Reconciler: &reconcileDeployAgentManifestwork{
+			client:        mgr.GetClient(),
+			addonLister:   addonLister,
+			clusterLister: clusterLister,
+			workClient:    workClient,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &addonv1alpha1.ManagedClusterAddOn{}}, &managedcluteraddonHandler{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *reconcileDeployAgentManifestwork) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// get managedClusterAddon name and managedClusterNamee
+	managedClusterAddonName := request.Name
+	managedClusterName := request.Namespace
+	if len(managedClusterName) == 0 || len(managedClusterAddonName) == 0 {
+		klog.Infof("Skip reconcile: managedClusterAddonName or managedClusterName is empty")
+		return reconcile.Result{}, nil
+	}
+
+	// get managedCluster
+	managedCluster, err := r.clusterLister.Get(managedClusterName)
+	if err != nil {
+		klog.Errorf("Failed to get managedCluster %q: %v", managedClusterName, err)
+		return reconcile.Result{}, err
+	}
+
+	// get managedClusterAddon
+	managedClusterAddon, err := r.addonLister.ManagedClusterAddOns(managedClusterName).Get(managedClusterAddonName)
+	if err != nil {
+		klog.Errorf("Failed to get managedClusterAddon %q: %v", managedClusterAddonName, err)
+		return reconcile.Result{}, err
+	}
+
+	// get server certificates
+	key, cert, err := getServerCertificatesFromSecret(r.client, certificatesNamespace)
+	if err != nil {
+		klog.Errorf("Failed to get server certificates: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	// get override image name
+	overrideImageName, err := getOverrideImageName(managedCluster, agentImage)
+	if err != nil {
+		klog.Errorf("Failed to get override image name: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	// get nodeSelector
+	nodeSelector, err := getNodeSelector(managedCluster)
```

Comment:
if non local-cluster, how do we handle this?

Reply:
If it's not a local-cluter, it wil be empty.
I think we need to handle addondeployment config also in this release with another PR to solve this.

Reply:
TODO added.